perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_user_attribute_def_virt_tcsMails_mu.java

```diff
@@ -0,0 +1,162 @@
+
+		Attribute attribute = new Attribute(attributeDefinition);
+		SortedSet<String> tcsMailsValue = new TreeSet<>();
+
+		tcsMailsValue.addAll(getEmailValues(sess, user, A_U_D_preferredMail));
+		tcsMailsValue.addAll(getEmailValues(sess, user, A_U_D_ISMail));
+		tcsMailsValue.addAll(getEmailValues(sess, user, A_U_D_o365EmailAddressesMU));
+		tcsMailsValue.addAll(getEmailValues(sess, user, A_U_D_publicAliasMails));
+		tcsMailsValue.addAll(getEmailValues(sess, user, A_U_D_privateAliasMails));
+
+		attribute.setValue(new ArrayList(tcsMailsValue));
+
+		return attribute;
+	}
+
+	@Override
+	public List<AuditEvent> resolveVirtualAttributeValueChange(PerunSessionImpl perunSession, AuditEvent message) throws InternalErrorException, WrongReferenceAttributeValueException, AttributeNotExistsException, WrongAttributeAssignmentException {
+		List<AuditEvent> resolvingMessages = new ArrayList<>();
+
+		// handle source user attributes changes
+		if (message instanceof AttributeSetForUser && isAffectedAttribute(((AttributeSetForUser) message).getAttribute().getFriendlyName())) {
+			AttributeDefinition attributeDefinition = perunSession.getPerunBl().getAttributesManagerBl().getAttributeDefinition(perunSession, this.getAttributeDefinition().getName());
+			resolvingMessages.add(new AttributeChangedForUser(new Attribute(attributeDefinition), ((AttributeSetForUser) message).getUser()));
+
+		} else if (message instanceof AttributeRemovedForUser && isAffectedAttribute(((AttributeRemovedForUser) message).getAttribute().getFriendlyName())) {
+			AttributeDefinition attributeDefinition = perunSession.getPerunBl().getAttributesManagerBl().getAttributeDefinition(perunSession, this.getAttributeDefinition().getName());
+			resolvingMessages.add(new AttributeChangedForUser(new Attribute(attributeDefinition), ((AttributeRemovedForUser) message).getUser()));
+
+		} else if (message instanceof AllAttributesRemovedForUser) {
+			AttributeDefinition attributeDefinition = perunSession.getPerunBl().getAttributesManagerBl().getAttributeDefinition(perunSession, this.getAttributeDefinition().getName());
+			resolvingMessages.add(new AttributeChangedForUser(new Attribute(attributeDefinition), ((AllAttributesRemovedForUser) message).getUser()));
+
+		}
+
+		return resolvingMessages;
+	}
+
+	/**
+	 * Return true if attribute name is one of affected attributes in this virtual module.
+	 *
+	 * @param nameOfAttribute name of attribute to check
+	 * @return true if attribute is one of affected, false otherwise
+	 */
+	private boolean isAffectedAttribute(String nameOfAttribute) {
+		if(preferredMailFriendlyName.equals(nameOfAttribute)) return true;
+		else if(isMailFriendlyName.equals(nameOfAttribute)) return true;
+		else if(o365MailsFriendlyName.equals(nameOfAttribute)) return true;
+		else if(publicMailsFriendlyName.equals(nameOfAttribute)) return true;
+		else if(privateMailsFriendlyName.equals(nameOfAttribute)) return true;
```

Comment:
You can change the last "else if" to: 
```
return (privateMailsFriendlyName.equals(nameOfAttribute));
```
Therefore, you do not have to have that extra code at the end of the method.

Reply:
I don't think it is necessary and it is easier to read the simple version than the shorter one.

Reply:
Ok :)